src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs

```diff
@@ -952,14 +951,28 @@ public static OperationStatus TranscodeToUtf8(char* pInputBuffer, int inputLengt
                         for (i = 0; (uint)i < maxIters; i++)
                         {
                             utf16Data = Unsafe.ReadUnaligned<Vector128<short>>(pInputBuffer);
-                            if (!Sse41.TestZ(utf16Data, nonAsciiUtf16DataMask))
+
+                            if (AdvSimd.IsSupported)
                             {
-                                goto LoopTerminatedDueToNonAsciiDataInVectorLocal;
-                            }
+                                if (AdvSimd.CompareTest(utf16Data, nonAsciiUtf16DataMask).ToScalar() == 0)
```

Comment:
Isn't this check backwards?

`bool TestZ` basically does `return (left & right) == 0`
`CompareTest` basically does `return (left & right) != 0 ? Vector128.AllBitsSet : Vector128.Zero`

Given the below check is `!TestZ` that means we are checking for `(left & right) != 0`, so `CompareTest` should also be `!= 0`, right?

Reply:
You're right, I got those two backwards. Thanks for correcting me.

- Sse41.TestZ: `Compute the bitwise AND of 128 bits (representing integer data) in a and b, and set ZF to 1 if the result is zero, otherwise set ZF to 0. Compute the bitwise NOT of a and then AND with b, and set CF to 1 if the result is zero, otherwise set CF to 0. Return the ZF value.`

- AdvSimd.CompareTest: `Compare bitwise Test bits nonzero (vector). This instruction reads each vector element in the first source SIMD&FP register, performs an AND with the corresponding vector element in the second source SIMD&FP register, and if the result is not zero, sets every bit of the corresponding vector element in the destination SIMD&FP register to one, otherwise sets every bit of the corresponding vector element in the destination SIMD&FP register to zero.`

Reply:
As pointed by @echesakovMSFT in https://github.com/dotnet/runtime/pull/39041#discussion_r453072911 , just `CompareTest()` is not good enough. After `CompareTest` you have 8 lanes, all of the bits in each lane are either 0 or 1. You need to further reduce it to bring it to 0th lane in order to do `.ToScalar()` on it.

Reply:
Or probably something like this: 
```csharp
Vector128<short> compareResult = AdvSimd.CompareTest(utf16Data, nonAsciiUtf16DataMask);
compareResult = AdvSimd.MinPairwise(compareResult, compareResult);
if (compareResult.AsUInt64().ToScalar() != 0) {
   goto LoopTerminatedDueToNonAsciiDataInVectorLocal;
}

```

Just double check it is functionally correct.